traceroute-eval/scripts/parse-traceroutes.py

```diff
@@ -0,0 +1,199 @@
+from os import listdir
+from os.path import isdir, join
+import json
+import re
+
+data_root = "../data"
+tr_methods = ["udp", "icmp", "tcp", "paris", "dublin"]
+output_file = "parse-traceroutes-output.json"
+
+# JSON keys
+TS = "Timestamp"
+DEST = "DestinationIP"
+TYPE = "TracerouteType"
+HOPS = "Hops"
+TTL = "TTL"
+IP = "IPs"
+RTT = "RTTs"
+
+# Regex patterns
+ip_pattern = re.compile(r"\d+\.\d+\.\d+\.\d+")
+ping_pattern = re.compile(r"\d+\.\d+")
+
+def base_parser(traceroute, method):
+    """
+    Parses one TCP, UDP, or, ICMP traceroute.
+
+    Args: 
+        traceroute (str): the full traceroute output
+        method (str)    : the type of traceroute
+    Returns:
+        json_dict (dict): dict containing relevant information
+    """
+
+    lines = traceroute.rstrip("\n").split("\n")
+
+    json_dict = {}
+    json_dict[TS] = lines[0]
+    json_dict[DEST] = lines[1].split()[2]
+    json_dict[TYPE] = method
+    json_dict[HOPS] = []
+
+    # Each line represents one hop
+    for hop in lines[2:-1]:
+        tokens = hop.split()
+        ttl = tokens[0]
+        ips = []
+        rtts = []
+        
+        # A token is either an IP, latency, "ms", or "*"
+        for token in tokens[1:]:
+            if token == "*":
+                ips.append("*")
+                rtts.append("*")
+            elif ip_pattern.match(token):
+                ips.append(token)
+            elif ping_pattern.match(token):
+                rtts.append(token)
+                # If same IP reached
+                if len(rtts) > len(ips):
```

Comment:
I'm not sure I understand what you mean here by "If same IP reached" adn then check for len rtt and len ips.

Reply:
So this is when we have something like `line = "[ip] [rtt1] [rtt2] [rtt3]"`. In this case, since one IP corresponds with all three `rtts`, we would just insert the same `ip` into the `ips` array three times.

A more complicated case which this also handles is `line = "[ip] [rtt1] * [rtt3]"`